include/boost/math/tools/polynomial.hpp

```diff
@@ -644,6 +645,30 @@ bool even(polynomial<T> const &a)
     return !odd(a);
 }
 
+template <class T>
+polynomial<T> pow(polynomial<T> base, int exp)
+{
+    if (exp < 0)
+        return policies::raise_domain_error(
+                "boost::math::tools::pow<%1%>",
+                "Negative powers are not supported for polynomials.",
+                base, policies::policy<>());
+        // if the policy is ignore_error or errno_on_error, raise_domain_error
+        // will return std::numeric_limits<polynomial<T>>::quiet_NaN(), which
+        // defaults to polynomial<T>(), which is the zero polynomial
+    polynomial<T> result(T(1));
+    if (exp & 1)
+        result = base;
+    /* "Exponentiation by squaring" */
+    while (exp >>= 1)
+    {
+        base *= base;
+        if (exp & 1)
+            result *= base;
+    }
+    return result;
```

Comment:
Apart from the constraint on `exp` this is after all a generic exponentiation algorithm, which would be preferable to see implemented separately from the context of the polynomial class.

Marshall actually implemented [one version](https://github.com/boostorg/algorithm/blob/develop/include/boost/algorithm/algorithm.hpp) (from Knuth) of this algorithm but it explicitly uses modulo to test for oddness, which is of course a performance drag and a reason why I think `odd` deserves to be a trait function for types.